tests/expectations.json

```diff
@@ -9213,97 +9213,8 @@
   "built-ins/WeakRef/proto-from-ctor-realm.js": "FAIL",
   "built-ins/WeakRef/prototype/deref/this-does-not-have-internal-target-throws.js": "FAIL",
   "built-ins/WeakSet/proto-from-ctor-realm.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.10_T1.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.11_T1.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.11_T2.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.12_T1.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.12_T2.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.12_T3.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.13_T1.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.13_T2.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.14_T1.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.14_T2.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.14_T3.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.14_T4.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.15_T1.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.15_T2.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.15_T3.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.15_T4.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.15_T5.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.15_T6.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.1_T1.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.2_T1.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.2_T2.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.3_T1.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.3_T2.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.4_T1.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.5_T1.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.6_T1.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.7_T1.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.8_T1.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.8_T2.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.9_T1.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.9_T2.js": "FAIL",
-  "built-ins/decodeURI/S15.1.3.1_A1.9_T3.js": "FAIL",
   "built-ins/decodeURI/S15.1.3.1_A2.1_T1.js": "FAIL",
```

Comment:
The remaining two tests fail because of https://github.com/trynova/nova/blob/main/small_string/lib.rs#L57

Reply:
I'll take a look.

Reply:
Alright, it was a fairly simple case of using the `string.len(agent)` function, ie. getting the UTF-8 length, but then using `string.utf16_char(agent, index)` function to get a char which expects a UTF-16 index.

I refactored the code a little to use more UTF-8 instead of unnecessarily converting into UTF-16 quite so much. Probably some of the code is now more incorrect than it was, but none of the tests fail so I guess it's good.